File: 146.LRU_Cache/LRU.js
class ListNode {
  constructor(key, value) {
    this.key = key;
    this.value = value;
    this.prev = null;
    this.next = null;
  }
}

/**
 * @param {number} capacity
 */
var LRUCache = function(capacity) {
  this.cap = capacity;
  this.map = {};
  this.head = new ListNode(0, 0);
  this.tail = new ListNode(0, 0);
  this.head.next = this.tail;
  this.tail.prev = this.head;
};

/**
 * @param {number} key
 * @return {number}
 */
LRUCache.prototype.get = function(key) {
  if (key in this.map) {
    node = this.map[key];
    this._remove(key);
    this._addToHead(key, node.value);
    return node.value;
  } else {
    return -1;
  }
};

/**
 * @param {number} key
 * @param {number} value
 * @return {void}
 */
LRUCache.prototype.put = function(key, value) {
  // if existed, change the value
  if (this.get(key) !== -1) {
    let node = this.map[key];
    node.value = value;
  } else {
    // reach the cap, remove the first element
    if (Object.keys(this.map).length === this.cap) {
      let firstNode = this.head.next;
      this._remove(firstNode.key);
    }
    this._addToHead(key, value);
  }
};

LRUCache.prototype._remove = function(key) {
  let node = this.map[key];
  let prev = node.prev;
  let next = node.next;
  if (prev !== null) prev.next = next;
  if (next !== null) next.prev = prev;
  delete this.map[key];
};

LRUCache.prototype._addToHead = function(key, value) {
  let newNode = new ListNode(key, value);
  let last = this.tail.prev;
  last.next = newNode;
  newNode.prev = last;
  newNode.next = this.tail;
  this.tail.prev = newNode;
  this.map[key] = newNode;
};
/**
 * Your LRUCache object will be instantiated and called as such:
 * var obj = new LRUCache(capacity)
 * var param_1 = obj.get(key)
 * obj.put(key,value)
//  */
// var obj = new LRUCache(2);
// var param_1 = obj.get(1);
// obj.put(1, 1);
// obj.put(2, 2);
// obj.put(3, 3);
// console.log(obj);

File: 146.LRU_Cache/lru.go
package LRU_Cache

type Node struct {
	// used for map to track
	key   int
	value int
	prev  *Node
	next  *Node
}

// keyMap here is used to reduce find complexity to O(1), and track the length of list
type LRUCache struct {
	capacity int
	head     *Node
	tail     *Node
	keyMap   map[int]*Node
}

func (this *LRUCache) add(n *Node) {
	last := this.tail.prev
	last.next = n
	this.tail.prev = n
	// points this node to the last and tail
	n.prev = last
	n.next = this.tail
}

func (this *LRUCache) remove(n *Node) {
	prev := n.prev
	next := n.next
	// modify pointer from last and tail to this node
	if n.prev != nil {
		n.prev.next = next
	}
	if n.next != nil {
		n.next.prev = prev
	}
}

func Constructor(capacity int) LRUCache {
	var cache LRUCache
	cache.capacity = capacity
	cache.head.next = cache.tail
	cache.tail.prev = cache.head
	return cache
}

func (this *LRUCache) Get(key int) int {
	if node, ok := this.keyMap[key]; ok {
		// put the node at the end of the list
		this.remove(node)
		this.add(node)
		return node.value
	}
	return -1
}

func (this *LRUCache) Put(key int, value int) {
	_, ok := this.keyMap[key]
	if ok != false {
		delete(this.keyMap, key)
	}
	n := Node{key: key, value: value}
	this.add(&n)
	this.keyMap[key] = &n
	if len(this.keyMap) > this.capacity {
		firstNode := this.head.next
		this.remove(firstNode)
		delete(this.keyMap, firstNode.key)
	}
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * obj := Constructor(capacity);
 * param_1 := obj.Get(key);
 * obj.Put(key,value);
 */
